p2p/wdrr_queue.go

```diff
@@ -0,0 +1,301 @@
+package p2p
+
+import (
+	"github.com/gogo/protobuf/proto"
+	"github.com/tendermint/tendermint/libs/log"
+	tmsync "github.com/tendermint/tendermint/libs/sync"
+)
+
+type drrFlow struct {
+	mtx    tmsync.RWMutex
+	size   uint
+	buffer []Envelope
+
+	capacity uint
```

Comment:
I'd set a capacity for the overall queue, not each flow. If e.g. 1 flow is full but all others are empty there's no need to start dropping messages. The sole purpose of capacity is to bound memory usage, and we should bound this per peer not per flow.